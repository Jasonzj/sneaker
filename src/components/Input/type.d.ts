import React from 'react'
import { DeepMap, FieldError, ValidationRules } from 'react-hook-form'

export type RefReturn =
  | string
  | ((instance: HTMLInputElement | null) => void)
  | React.RefObject<HTMLInputElement>
  | null
  | undefined

export type Props = React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement> & {
  type: string
  name: string
  placeholder: string
  rules: ValidationRules
  register: (rules: ValidationRules) => RefReturn
  errors: DeepMap<Record<string, unknown>, FieldError>
}
