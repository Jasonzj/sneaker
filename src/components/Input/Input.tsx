import React from 'react'
import classNames from 'classnames'
import ErrorMessage from '../../components/ErrorMessage'
import style from './input.module.css'
import { Props } from './type'

const Input: React.FC<Props> = ({ type, name, register, rules, errors, placeholder }) => (
  <div className={style.inputBox}>
    <input
      type={type}
      name={name}
      ref={register(rules)}
      placeholder={placeholder}
      className={classNames(style.input, {
        [style.error]: errors[name],
      })}
    />
    <ErrorMessage
      error={errors}
      name={name}
      render={(message) => <p className={style.errorMsg}>{message}</p>}
    />
  </div>
)

export default Input
