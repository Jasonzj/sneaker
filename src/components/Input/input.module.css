.inputBox {
  height: 70px;
}

.input {
  width: 100%;
  font-size: 14px;
  line-height: 1.5;
  display: block;
  box-sizing: border-box;
  border-radius: 4px;
  padding: 6px 10px;
  border: 1px solid #d9d9d9;
  color: #000;
  background-color: #fff;
  transition: all 0.3s;
}

.input:hover,
.input:focus {
  border-color: #49a9ee;
}

.input.error {
  border-width: 1px 1px 1px 10px;
  border-style: solid;
  border-color: #bf1650#bf1650#bf1650 #ec5990;
  border-image: initial;
  background: rgb(251, 236, 242);
}

.errorMsg {
  font-size: 14px;
  margin: 0;
  padding: 7px 0;
  color: #bf1650;
}

.errorMsg::before {
  display: inline;
  content: 'âš  ';
}
