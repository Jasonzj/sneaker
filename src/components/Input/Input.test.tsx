import { render, screen } from '@testing-library/react'
import Input from './'

const props = {
  type: 'text',
  name: 'username',
  register: jest.fn(),
  errors: {},
  rules: {},
  placeholder: 'username',
}

test('should render correctly placeholder and type', () => {
  const { asFragment } = render(<Input {...props} />)
  const input = screen.getByRole('textbox')

  expect(input).toHaveProperty('placeholder', 'username')
  expect(input).toHaveProperty('type', 'text')
  expect(asFragment()).toMatchSnapshot()
})

test('should render correctly className', () => {
  const { rerender, asFragment } = render(<Input {...props} />)
  const input = screen.getByRole('textbox')

  expect(input).not.toHaveClass('error')

  rerender(<Input {...props} errors={{ username: { type: 'username', message: 'error' } }} />)

  expect(input).toHaveClass('error')
  expect(asFragment()).toMatchSnapshot()
})
