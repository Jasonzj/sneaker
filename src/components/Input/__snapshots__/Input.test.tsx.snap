// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly className 1`] = `
<DocumentFragment>
  <div
    class="inputBox"
  >
    <input
      class="input error"
      name="username"
      placeholder="username"
      type="text"
    />
    <p
      class="errorMsg"
    >
      error
    </p>
  </div>
</DocumentFragment>
`;

exports[`should render correctly placeholder and type 1`] = `
<DocumentFragment>
  <div
    class="inputBox"
  >
    <input
      class="input"
      name="username"
      placeholder="username"
      type="text"
    />
  </div>
</DocumentFragment>
`;
